feat(android/app): Add option to change display language

darcywong00 opened the pull request:
Addresses #4236 and implements the option to set the app to Obolo from the strings in #4256

Currently, the Android ROM determines what locales are available to use. This PR adds a Keyman Settings option to change the display language (e.g. a "custom" language like Obolo `ann-latn`). The language is saved as a new shared preference and used on the next app startup.

[Reference tutorial](https://lokalise.com/blog/android-app-localization/#Change_Application_Locale_Programmatically)

## User Testing and Screenshots
@MakaraSok - See if you can reproduce these 

### Device Locale set to Khmer (App displaying default locale)
Generated by setting the device's primary language to "Khmer" and starting Keyman for Android (on Android Studio emulator, may need an Android version API 25+). 
With the device in Khmer locale, the Keyman app will "default" to Khmer
Note: the English strings in the screens are new from this PR that aren't on crowdin yet.

Start Keyman for Android and verify you see the following screens:

Getting Started
![khmer get started](https://user-images.githubusercontent.com/7358010/104668514-d03ed400-570a-11eb-98b2-9722ab98de7a.png)

Keyman Settings
![khmer settings](https://user-images.githubusercontent.com/7358010/104668526-d634b500-570a-11eb-872a-06ca5bad27af.png)

Change Display Language
Go to "Settings" ---> "Change Display Language" --> set to "Obolo"
![khmer set display language](https://user-images.githubusercontent.com/7358010/104668627-0a0fda80-570b-11eb-9649-f4d98aa8b772.png)

Changing the app's display language requires closing the app and restarting it

### Device locale set to "Khmer" (App display language "Obolo")
After setting Keyman's display language to "Obolo" and restarting the app
Verify you see the following screens

Getting Started
![obolo getting started](https://user-images.githubusercontent.com/7358010/104878491-a39aee80-598e-11eb-9e3f-432b696e3cf2.png)

Menus
![obolo menus](https://user-images.githubusercontent.com/7358010/104878503-a85fa280-598e-11eb-916e-4009fa0c68ce.png)

Settings
![obolo settings](https://user-images.githubusercontent.com/7358010/104878512-abf32980-598e-11eb-85c2-c38ebc532e3e.png)

Download keyboard package
"Settings" --> "Install from keyman.com" --> Search for sil_cameroon_qwerty and install.
You'll need to remember the row order in the "Settings" menu since we can't read Obolo :)

![obolo downloading](https://user-images.githubusercontent.com/7358010/104884791-86b7e880-5999-11eb-8d36-c140fd3bb58d.png)


Installing keyboard package
![obolo readme](https://user-images.githubusercontent.com/7358010/104884804-8cadc980-5999-11eb-9e8f-7960a65d5f92.png)

![obolo select language](https://user-images.githubusercontent.com/7358010/104884815-90d9e700-5999-11eb-89e3-e852390df370.png)

![obolo welcome](https://user-images.githubusercontent.com/7358010/104884828-946d6e00-5999-11eb-8222-51574e2cb084.png)

Keyboard Info (KMEA)
From the Keyman app, click globe key --> click the "i" info button on Cameroon QWERTY

![obolo keyboard info](https://user-images.githubusercontent.com/7358010/104884859-a0593000-5999-11eb-9e9b-c0a50e5c64df.png)


----

Still some TODO's before this is ready
- [x] UI is only changing for MainActivity 😞 
- [x] BaseActivity depends on the app's shared preference. I'm not sure how to get KMEA activities to change
- [ ] ~Setting Obolo language works on Android API 28, but not API 25~ Hopefully, not many users are still on API 25...

mcdurdin commented:
Good work. A couple of UI notes:
* I like the use of the language icon to make it easy to find the Display Language option no matter which language is selected.
* bug: "touch the globe" caption is still in Khmer when Obolo is selected in the screenshots above.
* nit: The "+" icon is not really needed in the list of display languages, and is it possible to vertically centre the language names in each row (it looks a little unbalanced otherwise)

mcdurdin commented on android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/KeymanSettingsActivity.java:
```diff
@@ -20,6 +22,7 @@ protected void onCreate(Bundle savedInstanceState) {
     setContentView(R.layout.keyman_settings_layout);
     Toolbar toolbar = (Toolbar) findViewById(R.id.keyman_settings_toolbar);
     toolbar.setTitleTextColor(getResources().getColor(android.R.color.white));
+    toolbar.setTitle(R.string.keyman_settings);
```
I take it this was a string we missed earlier?

darcywong00 replied:
Correct. When the display language was set to Obolo, I noticed the "Settings" menu still showed in Khmer.

mcdurdin commented on android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/KeymanSettingsLocalizeActivity.java:
```diff
@@ -0,0 +1,129 @@
+    int[] to = new int[]{com.tavultesoft.kmea.R.id.text1, com.tavultesoft.kmea.R.id.text2, com.tavultesoft.kmea.R.id.image1};
+
+    ListAdapter adapter = new SimpleAdapter(context, localizeOptionList, com.tavultesoft.kmea.R.layout.list_row_layout2, from, to) {
+      /*
+      @Override
+      public boolean isEnabled(int position) {
+        HashMap<String, String> hashMap = localizeOptionList.get(position);
+        String itemTitle = MapCompat.getOrDefault(hashMap, titleKey, "");
+        String icon = MapCompat.getOrDefault(hashMap, iconKey, noIcon);
+        if (itemTitle.equals(getString(R.string.install_from_other_device))) {
+          // Scan QR code not implemented yet
+          return false;
+        } else if (itemTitle.equals(getString(R.string.add_languages_to_installed_keyboard))) {
+          if (KeyboardController.getInstance().getInstalledPackagesList() == null) {
+            // Disable if no keyboard packages installed
+            return false;
+          }
+        }
+
+        return super.isEnabled(position);
+      }
+       */
```
Please remove commented code 😸

mcdurdin commented on android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/KeymanSettingsLocalizeActivity.java:
```diff
@@ -0,0 +1,129 @@
+
+      @Override
+      public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
+        HashMap<String, String> hashMap = (HashMap<String, String>) parent.getItemAtPosition(position);
```
This is not used?

mcdurdin commented on android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/KeymanSettingsLocalizeActivity.java:
```diff
@@ -0,0 +1,129 @@
+          SharedPreferences.Editor editor = prefs.edit();
+          editor.putString(DisplayLanguages.displayLanguageKey, languageTag);
+          editor.commit();
+          finish();
```
How does the user know they need to restart the app in order for the language to be reset? Is it possible to use [Activity.recreate](https://developer.android.com/reference/android/app/Activity.html#recreate%28%29) to avoid this user interaction?

darcywong00 replied:
I've updated the activity to restart the app.

mcdurdin commented on android/KMEA/app/src/main/java/com/tavultesoft/kmea/DisplayLanguages.java:
```diff
@@ -0,0 +1,39 @@
+
+  // Display Language types (as named in translate.keyman.com)
+  // Order doesn't matter since we're storing BCP-47 tags in the settings preference
+  public static final DisplayLanguageType[] DisplayLanguages =
+  {
+    new DisplayLanguageType(unspecifiedLocale, "Default Locale"),
+    new DisplayLanguageType("km-KH", "Khmer"),
+    new DisplayLanguageType("ann", "Obolo")
+  };
```
I take it this will need to be maintained manually? We need to document this as I suspect it will be forgotten otherwise!

darcywong00 replied:
Yes, this is maintained manually. I've added a note in resources/build/l10n/README.md

mcdurdin reviewed (changes requested):
I like this. Most of my comments are just tidyup things. Just would be good if we can find a way to recreate the main activity to avoid the user having to do it manually -- if I understand your comments correctly.

mcdurdin commented:
> * Hopefully, not many users are still on API 25...

We have roughly 1200 active users on API 25 and earlier.

darcywong00 commented:
> * bug: "touch the globe" caption is still in Khmer when Obolo is selected in the screenshots above.

Hmm, setting the locale in a WebView might be trickier. From this thread, it sounds like the WebView is handled by Chrome and uses the locales that Chrome has.

https://stackoverflow.com/questions/40398528/android-webview-language-changes-abruptly-on-android-7-0-and-above


> * and is it possible to vertically centre the language names in each row (it looks a little unbalanced otherwise)

It looks unbalanced because we're not displaying subtext below the language names.  Same layout as adding a language to an installed keyboard
![select language](https://user-images.githubusercontent.com/7358010/104982747-da7c0d80-5a3d-11eb-8664-b07c4e359957.jpg)

mcdurdin commented on android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/KeymanSettingsLocalizeActivity.java:
```diff
@@ -0,0 +1,106 @@
+          editor.putString(DisplayLanguages.displayLanguageKey, languageTag);
+          editor.commit();
+
+          // Restart the app to use selected locale
+          Intent intent = new Intent(context, MainActivity.class);
+          intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
+          startActivity(intent);
```
👏 🥳

mcdurdin reviewed (commented):
> > * bug: "touch the globe" caption is still in Khmer when Obolo is selected in the screenshots above.
> 
> Hmm, setting the locale in a WebView might be trickier. From this thread, it sounds like the WebView is handled by Chrome and uses the locales that Chrome has.

We should be able to bypass that and pass the appropriate value in, surely? That little touch the globe caption is set by us.

> It looks unbalanced because we're not displaying subtext below the language names. Same layout as adding a language to an installed keyboard

Is it not possible to have the text vertically centred when we have no subtext?

darcywong00 commented:
> Is it not possible to have the text vertically centred when we have no subtext?

Ah - I changed from `list_row_layout2` to `list_row_layout1` to center each row of text

![display languages](https://user-images.githubusercontent.com/7358010/105119106-38722900-5b02-11eb-8996-3f942b326601.png)

Also added German locale from #4295 
![german menu](https://user-images.githubusercontent.com/7358010/105119119-3f00a080-5b02-11eb-8f91-566616bdab7b.png)

darcywong00 commented:
> We should be able to bypass that and pass the appropriate value in, surely? That little touch the globe caption is set by us.

I wonder if the locale is cached. I notice if I go into the Settings menu and return to the app, the globe is in the correct locale.

MakaraSok commented:
@darcywong00 Attested. Everything is shown as expected.

mcdurdin reviewed (approved):
LGTM. The failing Windows build is not relevant (that comes from a broken beta build which Joshua is dealing with).

keyman-server commented:
Changes in this pull request will be available for download in [Keyman version 14.0.226-beta](https://keyman.com/downloads/releases/beta/14.0.226)
